flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/DeclarativeSlotManager.java

```diff
@@ -407,13 +420,39 @@ public void freeSlot(SlotID slotId, AllocationID allocationId) {
         LOG.debug("Freeing slot {}.", slotId);
 
         slotTracker.notifyFree(slotId);
-        checkResourceRequirements();
+        checkResourceRequirementsWithDelay();
     }
 
     // ---------------------------------------------------------------------------------------------
     // Requirement matching
     // ---------------------------------------------------------------------------------------------
 
+    /**
+     * Depending on the implementation of {@link ResourceAllocationStrategy}, checking resource
+     * requirements and potentially making a re-allocation can be heavy. In order to cover more
+     * changes with each check, thus reduce the frequency of unnecessary re-allocations, the checks
+     * are performed with a slight delay.
+     */
+    private void checkResourceRequirementsWithDelay() {
+        if (requirementsCheckDelay.toMillis() <= 0) {
```

Comment:
If we allow users to configure the delay period, we should also check it in `FineGrainedSlotManager`

Reply:
Yes, will fix it